packages/worker/src/boot/startWorker.ts

```diff
@@ -1,76 +1,104 @@
+      try {
+        const response = handleAction(streams, event.data)
+        if (response) {
+          self.postMessage(response)
         }
-      })
-    )
+      } catch (error) {
+        sendError(error, event.data && 'streamId' in event.data ? event.data.streamId : undefined)
+      }
+    })
+  } catch (error) {
+    sendError(error)
+  }
+}
+
+function sendError(error: unknown, streamId?: number) {
+  try {
+    self.postMessage({
+      type: 'errored',
+      error,
+      streamId,
+    })
+  } catch (_) {
+    // DATA_CLONE_ERR, cf https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm
+    self.postMessage({
+      type: 'errored',
+      error: String(error),
+      streamId,
+    })
+  }
+}
+
+function handleAction(streams: Map<number, Deflate>, message: DeflateWorkerAction): DeflateWorkerResponse | undefined {
+  switch (message.action) {
+    case 'init':
+      return {
+        type: 'initialized',
+        version: __BUILD_ENV__SDK_VERSION__,
+      }
+
+    case 'write': {
+      let deflate = streams.get(message.streamId)
+      if (!deflate) {
+        deflate = new Deflate()
+        streams.set(message.streamId, deflate)
+      }
+      const previousChunksLength = deflate.chunks.length
 
-    function pushData(data: string) {
       // TextEncoder is not supported on old browser version like Edge 18, therefore we use string2buf
-      const binaryData = string2buf(data)
+      const binaryData = string2buf(message.data)
       deflate.push(binaryData, constants.Z_SYNC_FLUSH)
-      rawBytesCount += binaryData.length
-      return binaryData.length
-    }
-  })()
-}
 
-function monitor<Args extends any[], Result>(fn: (...args: Args) => Result): (...args: Args) => Result | undefined {
-  return (...args) => {
-    try {
-      return fn(...args)
-    } catch (e) {
-      try {
-        self.postMessage({
-          type: 'errored',
-          error: e,
-        })
-      } catch (_) {
-        // DATA_CLONE_ERR, cf https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm
-        self.postMessage({
-          type: 'errored',
-          error: String(e),
-        })
+      return {
+        type: 'wrote',
+        id: message.id,
+        streamId: message.streamId,
+        result: concatBuffers(deflate.chunks.slice(previousChunksLength)),
+        trailer: makeTrailer(deflate),
+        additionalBytesCount: binaryData.length,
       }
     }
+
+    case 'reset':
+      streams.delete(message.streamId)
+      break
   }
 }
+
+function concatBuffers(buffers: Uint8Array[]) {
+  const length = buffers.reduce((total, buffer) => total + buffer.length, 0)
+  const result = new Uint8Array(length)
+  let offset = 0
+  for (const buffer of buffers) {
+    result.set(buffer, offset)
+    offset += buffer.length
+  }
+  return result
+}
+
+function makeTrailer(deflate: Deflate): Uint8Array {
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏explaining a bit what is the trailer could be helpful, FMU it is some kind of checksum that is added at the end of the stream, right?

Reply:
I added some doc